src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ *
+ * RTC block 1 has an unknown purpose and is typically unimplemented.
+ * RTC block 2 contains the date and time.
+ */
+#define JOYBUS_RTC_CONTROL_MODE_RUN 0x0300
+
+/**
+ * @brief Decode a binary-coded decimal number.
+ *
+ * @param[in]   bcd
+ *              binary-coded decimal number
+ *
+ * @return the decoded number
+ */
+static uint8_t bcd_to_byte(uint8_t bcd)
+{
+    uint8_t hi = (bcd & 0xF0) >> 4;
+    uint8_t lo = bcd & 0x0F;
+    return (hi * 10) + lo;
+}
+
+/**
+ * @brief Encode a binary-coded decimal number.
+ *
+ * @param[in]   byte
+ *              number to encode
+ *
+ * @return the encoded binary-coded decimal number
+ */
+static uint8_t byte_to_bcd(uint8_t byte)
+{
+    byte %= 100;
+    return ((byte / 10) << 4) | (byte % 10);
+}
+
+/**
+ * @brief Compare two RTC times for (rough) equality
+ *
+ * @param[in]   a
+ *              Source pointer to an RTC time data structure
+ *
+ * @param[in]   b
+ *              Source pointer to an RTC time data structure
+ *
+ * @return whether the RTC time data structures are (roughly) equal
+ */
+static bool joybus_rtc_time_equal( const rtc_time_t * a, const rtc_time_t * b )
+{
+    return (
+        a->year  == b->year  &&
+        a->month == b->month &&
+        a->day   == b->day   &&
+        a->hour  == b->hour  &&
+        a->min   == b->min
+        /* Skip sec comparison in case the clock ticked between RTC reads */
+    );
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ *
+ * The RTC status response contains an identifier byte to indicate that the
+ * RTC is present on the cartridge and a status byte to indicate whether the
+ * RTC is in the "busy" state.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_is_present and #joybus_rtc_is_busy
+ * functions which handle the magic numbers associated with the response.
+ *
+ * @return the RTC status response data
+ */
+uint16_t joybus_rtc_status( void )
+{
+    static const uint64_t input[8] =
+    {
+        0x00000000ff010306,
+        0xfffffffe00000000,
+        0,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    controller_exec_pif( input, output );
+
+    uint8_t * recv_bytes = (uint8_t *)&output[1];
+
+    return (uint16_t)recv_bytes[1] << 8 | recv_bytes[2];
+}
+
+/**
+ * @brief Read a block of data from the real-time clock
+ *
+ * @param[in]   block
+ *              Which RTC block to read from (0-2)
+ *
+ * @param[out]  data
+ *              Destination pointer for the RTC block data
+ *
+ * @return the busy status of the real-time clock
+ */
+uint8_t joybus_rtc_read( uint8_t block, uint64_t * data )
+{
+    assert(block <= 2);
+
+    static uint64_t input[8] =
+    {
+        0x0000000002090700,
+        0xffffffffffffffff,
+        0xfffe000000000000,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
```

Comment:
Why this has to be static?

Reply:
I copied this pattern from `controller.c` -- I have no other reasoning for it being `static`

Reply:
Oh I see. I think it's not required (or I can't see how it should be required). Since it's already difficult to check this PR, maybe we should remove it from controller.c first and check it there.

Reply:
Fixed